### PR TITLE
FINERACT-1926: Asset Externalization on Fineract

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

API for Asset Externalization Feature, in order to  analyze and work on the asset external owner details, I need to be able to view filer and search  the AEO 

[FINERACT-1926](https://issues.apache.org/jira/browse/FINERACT-1926)

Screenshot 2023-08-03 at 21 14 55

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by adamsaghy** [comment]
@josehernandezfintecheandomx Please add proper FINERACT story number into the PR title and please amend your commit as well as that one should contain the relevant FINERACT story number as well!

File: fineract-investor/src/main/java/org/apache/fineract/investor/api/search/ExternalAssetOwnersSearchApiResource.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.fineract.investor.service.search.domain.ExternalAssetOwnerSearchData;
+import org.apache.fineract.investor.service.search.domain.ExternalAssetOwnerSearchRequest;
+import org.springframework.data.domain.Page;
+import org.springframework.stereotype.Component;
+
+@Path("/v1/external-asset-owner")
+@Component
+@Tag(name = "ExternalAssetOwnerSearchV1")
+@RequiredArgsConstructor
+public class ExternalAssetOwnersSearchApiResource implements ExternalAssetOwnersSearchApi {
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/api/search/ExternalAssetOwnersSearchApiResource.java]: i dont think it requires its own API class! this search endpoint can be moved into the `ExternalAssetOwnersApiResource`.

Also we are already supporting the `/v1/external-asset-owners` endpoint, we should not have a `/v1/external-asset-owner`

File: fineract-investor/src/main/java/org/apache/fineract/investor/domain/search/SearchedExternalAssetOwner.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.apache.fineract.infrastructure.core.domain.ExternalId;
+import org.apache.fineract.investor.data.ExternalTransferStatus;
+
+@Getter
+@RequiredArgsConstructor
+public class SearchedExternalAssetOwner {
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/domain/search/SearchedExternalAssetOwner.java:31]: Please reuse the already existing `ExternalTransferData` data object for this purpose!

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    public Page<ExternalAssetOwnerSearchData> searchInvestorData(PagedRequest<ExternalAssetOwnerSearchRequest> searchRequest) {
+        validateTextSearchRequest(searchRequest);
+        return executeSearch(searchRequest);
+    }
+
+    private void validateTextSearchRequest(PagedRequest<ExternalAssetOwnerSearchRequest> searchRequest) {
+        Objects.requireNonNull(searchRequest, "searchRequest must not be null");
+
+        context.isAuthenticated();
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java]: please do the authentication in the API layer

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    private final SearchingExternalAssetOwnerRepository externalAssetOwnerRepository;
+    private final ExternalAssetOwnerSearchDataMapper externalAssetOwnerSearchDataMapper;
+
+    public Page<ExternalAssetOwnerSearchData> searchInvestorData(PagedRequest<ExternalAssetOwnerSearchRequest> searchRequest) {
+        validateTextSearchRequest(searchRequest);
+        return executeSearch(searchRequest);
+    }
+
+    private void validateTextSearchRequest(PagedRequest<ExternalAssetOwnerSearchRequest> searchRequest) {
+        Objects.requireNonNull(searchRequest, "searchRequest must not be null");
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java:46]: Please enhance this validation and check whether any field of the ExternalAssetOwnerSearchRequest was set! We should avoid the no filtering searches! For that there is the GET `/transfers` API

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+        context.isAuthenticated();
+    }
+
+    private Page<ExternalAssetOwnerSearchData> executeSearch(PagedRequest<ExternalAssetOwnerSearchRequest> searchRequest) {
+        Optional<ExternalAssetOwnerSearchRequest> request = searchRequest.getRequest();
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java]: is there a reason why you are fetching the data from `ExternalAssetOwnerSearchRequest` and creating a new `ExternalAssetOwnerSearchRequest` instance with the same values?

I think the `searchRequest` AS-IS can be a parameter of the `externalAssetOwnerRepository.searchInvestorData`. What do you think?

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/domain/ExternalAssetOwnerSearchData.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+package org.apache.fineract.investor.service.search.domain;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import lombok.Data;
+import org.apache.fineract.infrastructure.core.domain.ExternalId;
+import org.apache.fineract.investor.data.ExternalTransferStatus;
+
+@Data
+public class ExternalAssetOwnerSearchData {
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/service/search/domain/ExternalAssetOwnerSearchData.java]: Please reuse the already existing ExternalTransferData data object for this purpose!

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments, questions!
Also please cover this functionality with tests!

File: fineract-investor/src/main/java/org/apache/fineract/investor/api/search/ExternalAssetOwnersSearchApiResource.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-investor/src/main/java/org/apache/fineract/investor/api/search/ExternalAssetOwnersSearchApiResource.java]: Done! Added in the `ExternalAssetOwnersApiResource` class and removed the `ExternalAssetOwnersSearchApiResource`

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java]: Done! Using platformUserRightsContext

File: fineract-investor/src/main/java/org/apache/fineract/investor/domain/search/SearchedExternalAssetOwner.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply fineract-investor/src/main/java/org/apache/fineract/investor/domain/search/SearchedExternalAssetOwner.java:31]: Done! data object updated and generated using a Mapper, here is a sample of this:

Screenshot 2023-08-03 at 21 14 55

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java (hunk 4 above)
Comment (josehernandezfintecheandomx) [reply fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java:46]: As we talked, we keep this, considering to be present the first items loaded without any parameter

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java (hunk 5 above)
Comment (josehernandezfintecheandomx) [reply fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java]: Done!

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/domain/ExternalAssetOwnerSearchData.java (hunk 6 above)
Comment (josehernandezfintecheandomx) [reply fineract-investor/src/main/java/org/apache/fineract/investor/service/search/domain/ExternalAssetOwnerSearchData.java]: Done! data object updated and generated using a Mapper

**Comment by adamsaghy** [comment]
@josehernandezfintecheandomx please fix the compilation errors!
```
Execution failed for task ':fineract-investor:compileTestJava'.

> Compilation failed; see the compiler error output for details.

File: fineract-investor/src/main/java/org/apache/fineract/investor/domain/search/SearchedExternalAssetOwner.java (hunk 2 above)
Comment (adamsaghy) [reply fineract-investor/src/main/java/org/apache/fineract/investor/domain/search/SearchedExternalAssetOwner.java:31]: So this class can be removed?

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        Objects.requireNonNull(searchRequest, "searchRequest must not be null");
+
+        platformUserRightsContext.isAuthenticated();
+    }
+
+    private Page<ExternalTransferData> executeSearch(PagedRequest<ExternalAssetOwnerSearchRequest> searchRequest) {
+        Optional<ExternalAssetOwnerSearchRequest> request = searchRequest.getRequest();
+        Pageable pageable = searchRequest.toPageable();
+
+        return externalAssetOwnerRepository.searchInvestorData(request.get(), pageable).map(externalAssetOwnerSearchDataMapper::map);
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java:59]: you can just simply call the `externalAssetOwnerRepository.searchInvestorData` with the searchRequest.

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java (hunk 5 above)
Comment (adamsaghy) [reply fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java]: you can just simply call the `externalAssetOwnerRepository.searchInvestorData` with the searchRequest.

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/mapper/ExternalAssetOwnerSearchDataMapper.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    default ExternalTransferSubStatus toSubStatus(SearchedExternalAssetOwner source) {
+        return source.getSubStatus();
+    }
+
+    @Named("toDetails")
+    default ExternalTransferDataDetails toDetails(SearchedExternalAssetOwner source) {
+        return new ExternalTransferDataDetails(source.getDetailsId(), source.getTotalOutstanding(), source.getPrincipalOutstanding(),
+                source.getInterestOutstanding(), source.getFeeOutstanding(), source.getPenaltyOutstanding(), source.getTotalOverpaid());
+    }
+}
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/service/search/mapper/ExternalAssetOwnerSearchDataMapper.java:77]: Please add tests to cover these newly implemented functionalities and behaviour!

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments!

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/mapper/ExternalAssetOwnerSearchDataMapper.java (hunk 8 above)
Comment (josehernandezfintecheandomx) [reply fineract-investor/src/main/java/org/apache/fineract/investor/service/search/mapper/ExternalAssetOwnerSearchDataMapper.java:77]: Done

**Comment by adamsaghy** [comment]
@josehernandezfintecheandomx Please add FINERACT-XXXX story number into the PR and into the commit!

File: fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResourceSwagger.java [hunk 9]
```diff
@@ -152,4 +152,57 @@ static final class ExternalAssetOwnerTransferChangesData {
...
+        public String effectiveFromDate;
+
+        @Schema(example = "2023-06-23")
+        public String effectiveToDate;
+
+        @Schema(example = "1234567890987654321abc")
+        public String text;
+    }
+
+    @Schema(description = "GetSearchExternalAssetOwnerTransferResponse")
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResourceSwagger.java]: We dont need this. Swagger can automatically generate this. See `ExternalAssetOwnersApiResource.getTransfers`

File: fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResourceSwagger.java [hunk 10]
```diff
@@ -152,4 +152,57 @@ static final class ExternalAssetOwnerTransferChangesData {
             public String purchasePriceRatio;
         }
     }
+
+    @Schema(description = "GetExternalAssetOwnerSearchRequest")
+    public static final class GetExternalAssetOwnerSearchRequest {
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResourceSwagger.java]: This is not used. not needed

File: fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResourceSwagger.java [hunk 11]
```diff
@@ -28,59 +28,59 @@ final class ExternalAssetOwnersApiResourceSwagger {
...
 
-            @Schema(example = "PENDING")
-            public ExternalTransferStatus status;
+        @Schema(example = "[2023, 5, 23]")
+        public LocalDate effectiveTo;
+    }
 
-            @Schema(example = "[2023, 5, 1]")
-            public LocalDate effectiveFrom;
+    @Schema(description = "ExternalTransferResponse")
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResourceSwagger.java:80]: you might wanna revert this rearrangement changes.

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+    public Page<ExternalTransferData> searchInvestorData(PagedRequest<ExternalAssetOwnerSearchRequest> searchRequest) {
+        validateTextSearchRequest(searchRequest);
+        return executeSearch(searchRequest);
+    }
+
+    private void validateTextSearchRequest(PagedRequest<ExternalAssetOwnerSearchRequest> searchRequest) {
+        Objects.requireNonNull(searchRequest, "searchRequest must not be null");
+
+        platformUserRightsContext.isAuthenticated();
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java]: Please move this authentication check to API layer!

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/SearchExternalAssetOwnerTransferTest.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        Utils.initializeRESTAssured();
+        REQUEST_SPEC = new RequestSpecBuilder().setContentType(ContentType.JSON).build();
+        REQUEST_SPEC.header("Authorization", "Basic " + Utils.loginIntoServerAndGetBase64EncodedAuthenticationKey());
+        RESPONSE_SPEC = new ResponseSpecBuilder().expectStatusCode(200).build();
+        EXTERNAL_ASSET_OWNER_HELPER = new ExternalAssetOwnerHelper();
+    }
+
+    @Test
+    public void initialSearchExternalAssetOwnerTransferTest() {
+        PagedRequestExternalAssetOwnerSearchRequest searchRequest = EXTERNAL_ASSET_OWNER_HELPER.buildExternalAssetOwnerSearchRequest("",
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/SearchExternalAssetOwnerTransferTest.java]: Please write proper tests!
1. Test case that covers searching by text
1.1 Test case when there is found item
1.2 Test case when there is no found item
2. Test case that covers searching by submitted on from/to date
2.1 Test case when there is found item
2.2 Test case when there is no found item
3. Test case that covers searching by effective from/to date
3.1 Test case when there is found item
3.2 Test case when there is no found item
4. Test case that covers searching by text and submitted on from date when it should find item by text, but it filters out as the submitted on from date is not matching
5. Test case when submitted on from date of search request date format is invalid

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please see my comments!

File: fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResourceSwagger.java (hunk 10 above)
Comment (josehernandezfintecheandomx) [reply fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResourceSwagger.java]: Done! removed

**Review by josehernandezfintecheandomx — COMMENTED** [review]
Done!

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java (hunk 12 above)
Comment (josehernandezfintecheandomx) [reply fineract-investor/src/main/java/org/apache/fineract/investor/service/search/ExternalAssetOwnerSearchService.java]: DOne, !t was moved to the API layer

File: fineract-investor/src/main/java/org/apache/fineract/investor/service/search/mapper/ExternalAssetOwnerSearchDataMapper.java (hunk 8 above)
Comment (adamsaghy) [reply fineract-investor/src/main/java/org/apache/fineract/investor/service/search/mapper/ExternalAssetOwnerSearchDataMapper.java:77]: I am sorry, but i dont see any of the requested use cases...

**Comment by adamsaghy** [comment]
@josehernandezfintecheandomx Please let me know when all the comments are replied / resolved. Thanks! ;)

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/SearchExternalAssetOwnerTransferTest.java (hunk 13 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/SearchExternalAssetOwnerTransferTest.java]: Done the tc

Screenshot 2023-08-29 at 23 33 44

**Review by adamsaghy — CHANGES_REQUESTED** [review]
1. HttpMessageNotReadableErrorController is okay, however i reckon we need an erorr handler to handle DateTimeParseException and provide back a valid error message with a hint about the problem. Something like this:
HTTP 400 - Bad request
Error message: Invalid date was provided. The date must match with the “yyyy-MMMM-dd” date format.
Can you please write an ErrorMapper for the DateTimeParseException?

2. ExternalAssetOwnersApiResourceSwagger.java -> Please undo the changes, they are not required. , Spring will generate automatically the swagger

3. ExternalAssetOwnersApiResource.java -> searchInvestorData method @ApiResponses annotation can be removed as well

4. Please enhance your tests cases:
- assertNotNull(response); is not enough… Please write assertion for each value and check the value is matching with the expectations
- Test case when submitted on from date of search request date format is invalid -> I dont see this test case to be covered!

**Comment by josehernandezfintecheandomx** [comment]
> 1. HttpMessageNotReadableErrorController is okay, however i reckon we need an erorr handler to handle DateTimeParseException and provide back a valid error message with a hint about the problem. Something like this:
>    HTTP 400 - Bad request
>    Error message: Invalid date was provided. The date must match with the “yyyy-MMMM-dd” date format.
>    Can you please write an ErrorMapper for the DateTimeParseException?
> 2. ExternalAssetOwnersApiResourceSwagger.java -> Please undo the changes, they are not required. , Spring will generate automatically the swagger
> 3. ExternalAssetOwnersApiResource.java -> searchInvestorData method @ApiResponses annotation can be removed as well
> 4. Please enhance your tests cases:
> 
> * assertNotNull(response); is not enough… Please write assertion for each value and check the value is matching with the expectations
> * Test case when submitted on from date of search request date format is invalid -> I dont see this test case to be covered!

**Done those changes and PR updated please review it, Thanls**

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/ExternalAssetOwnerTransferTest.java [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+import org.apache.fineract.integrationtests.common.loans.LoanApplicationTestBuilder;
+import org.apache.fineract.integrationtests.common.loans.LoanProductTestBuilder;
+import org.apache.fineract.integrationtests.common.loans.LoanStatusChecker;
+import org.apache.fineract.integrationtests.common.loans.LoanTransactionHelper;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+
+@Slf4j
+public class ExternalAssetOwnerTransferTest {
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/investor/externalassetowner/ExternalAssetOwnerTransferTest.java:72]: This class have 0 tests... 

Probably the class name is not the best...

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**